lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,468 @@
+	switch len(cmd.Args()) {
+	case 1:
+		// Passed AUTH command without any arguments. Mimic Redis error response.
+		return errors.New("wrong number of arguments for 'auth' command")
+	case 2:
+		// Old Redis command. Password is the only argument here. Pass to Redis to validate.
+		// ex. AUTH my-secret-password
+		return redisClient.Process(ctx, cmd)
+	case 3:
+		// Redis 6 version that contains username and password. Check the username against our RBAC before sending to Redis.
+		// ex. AUTH bob my-secret-password
+		dbUser, ok := cmd.Args()[1].(string)
+		if !ok {
+			return errors.New("username has a wrong type, expected string")
+		}
+
+		err := e.sessionCtx.Checker.CheckAccess(e.sessionCtx.Database,
+			services.AccessMFAParams{Verified: true},
+			role.DatabaseRoleMatchers(
+				defaults.ProtocolRedis,
+				dbUser,
+				// pass empty database as Redis integration doesn't support db name validation.
+				"")...)
+		if err != nil {
+			return err
+		}
+
+		return redisClient.Process(ctx, cmd)
+	default:
+		// Redis returns "syntax error" if AUTH has that 2 arguments.
+		return errors.New("syntax error")
+	}
+}
+
+// process is the main processing function for Redis. It reads commands passed from client and passes them to
+// a Redis instance. It's also responsible for audit.
+func (e *Engine) process(ctx context.Context, redisClient redis.UniversalClient) error {
+	for {
+		// Read commands from client.
+		cmd, err := e.readClientCmd(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		e.Audit.OnQuery(e.Context, e.sessionCtx, common.Query{Query: cmd.String()})
+
+		e.Log.Debugf("redis cmd: %s", cmd.String())
+
+		// send valid commands to Redis instance/cluster.
+		err = e.processCmd(ctx, redisClient, cmd)
+
+		var vals interface{}
+		if _, ok := err.(redis.Error); ok {
+			vals = err
+		} else if errors.Is(err, context.DeadlineExceeded) {
+			// Do not return Deadline Exceeded to the client as it's not very self-explanatory.
+			// Return "connection timeout" as this is what most likely happened.
+			vals = errors.New("connection timeout")
+		} else if err != nil {
+			// Send the error to the client as connection will be terminated after return.
+			e.SendError(err)
+
+			return trace.Wrap(err)
+		} else {
+			vals, err = cmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		// Send response back to the client.
+		if err := e.sendToClient(vals); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// writeCmd writes Redis commands passed as vals to Redis wire form.
+// Most types is covered by go-redis implemented WriteArg() function. Types override by this function are:
+// * Redis errors and Go error: go-redis returns a "human-readable" string instead of RESP compatible error message
+// * integers: go-redis converts them to string, which is not always what we want to.
+// * slices: arrays are recursively converted to RESP responses.
+func writeCmd(wr *redis.Writer, vals interface{}) error {
```

Comment:
I'd consider replacing `*redis.Writer` with an interface combining `io.Writer`, `io.ByteWriter` and `io.StringWriter`. Would make implementation a little more flexible and easier to unit-test protocol.

Reply:
Sounds like a good idea, but then I lose access to a few `redis.Writer` specific methods. Let me think about it.